### PR TITLE
Masterbar: Refactor, decouple logged-in and out logic

### PR DESCRIPTION
**Pull request by ockham**
The aim of this PR is to untangle the logged-in and logged-out versions of the masterbar in order to later be able to render the logged-out layout without any unnecessary dependencies etc. See https://github.com/Automattic/wp-calypso/pull/1918#issuecomment-166904527 for a PoC.

Note that this PR also removes the `collapsible` prop and class from the masterbar, since the corresponding styling was removed in 355cd948d8ad2564fe15a92dd4740a9085f4d99b by @folletto.

Addresses #1326. No visible changes.

To test: Verifiy that the masterbar works in both 
 * logged-in and
 * logged-out sections. For the latter, open a browser window in incognito mode and navigate to either calypso.localhost:3000/start or calypso.localhost:3000/design

### CODE REVIEW

**Comment by ehg** [comment]
We should avoid a `package.json` split, as the tree should be exactly the same on client/server :)

**Comment by ockham** [comment]
There goes my plan. :,-(

File: client/layout/masterbar/index.jsx [hunk 1]
```diff
@@ -63,70 +22,11 @@ export default React.createClass( {
...
-			);
-		}
-
-		// Logged out
 		return (
 			<header id="header" className={ classes }>
-				<Item url="/" icon="my-sites" className="masterbar__item-logo">
-					WordPress<span className="tld">.com</span>
-				</Item>
+				{ this.props.user
```
Comment (ehg) [inline client/layout/masterbar/index.jsx]: I was thinking we'd just accept `children` here, and have the parent components care about rendering the loggedin/out components.

Comment (ockham) [inline client/layout/masterbar/index.jsx]: Makes sense, I'll change it like that.

**Comment by ockham** [comment]
Tentative `renderToString()` usage: https://gist.github.com/ockham/f74ae0b6b332124360f0

**Comment by ockham** [comment]
^ Currently failing because `components/global-notices` are in `client` and can't thus be found.

**Comment by ehg** [comment]
Updated SSR patch: https://gist.github.com/anonymous/5de3b7c46731331a9651

Use `curl https://gist.githubusercontent.com/anonymous/28201f911eee1ee5939c/raw/3bc1bc9df48d24afe16079232377e0fe9c1c3375/- | git apply` to apply it, may need a `git stash` first.

With it, the Masterbar renders cleanly on the server. Nice.

**Comment by ockham** [comment]
Glad to see this works already if we include components from `client/`.

So to keep this lean and mean, let's start wrapping this up. One question about directory structure -- where should everything go? Currently,

* `layout/masterbar/index.jsx` is just a wrapper component and hence doesn't depend on any of the other stuff in that directory
* `layout/masterbar/logged-out.jsx` dependes only on `layout/masterbar/item.jsx`
* `layout/masterbar/logged-in.jsx` depends on pretty much every other file in that dir, including `item.jsx`

`layout/logged-out.jsx` then requires `layout/masterbar/logged-out.jsx`, while `layout/index.jsx` requires `layout/masterbar/logged-in.jsx`. Both also use the `layout/masterbar/index.jsx` wrapper.

**Comment by ockham** [comment]
Also noting that I made a mistake -- `<header>` is still present in `logged-in.jsx`, so it's duplicated (since it's also in `index.jsx`). Alas, we can't just go ahead and remove it from `logged-in.jsx` since it has multiple children there, and we need one common root component :-/

**Comment by ockham** [comment]
...and just inserting a wrapping `<div>` randomly seems really bad, given that we might just as well keep `<header>` there, which is much more semantic.

So maybe revert 033a13284af4fe48e58b3343bf25ed73b3757ea6 after all, and live with `LoggedInMasterbar` and `LoggedOutMasterbar` being composed from `Masterbar`?

**Comment by ockham** [comment]
Or, alternatively, drop both, and just put `<Masterbar<`'s child components into `layout/logged-out.jsx` and `layout/index.jsx` directly.

`</StreamOfConsciousness>`

**Comment by ockham** [comment]
Nevermind, figured out the probably only sane (and pretty obvious) way to do this :-)

**Comment by mcsf** [comment]
Noting that I'm getting a:

```Warning: Failed propType: Invalid prop `children` supplied to `Masterbar`, expected a single ReactElement. Check the render method of `Masterbar`.```

**Comment by mcsf** [comment]
^ Potential fixes:
- Wrapping all the children of `Masterbar` in something neutral like a `div`
- Changing the type to `React.PropTypes.arrayOf( React.PropTypes.element )` and assuming multiple children will always be provided.
- Simply changing the type to `React.PropTypes.node`, which is more lax.

**Comment by mcsf** [comment]
I'll give this some more testing later, but it generally looks good! Quite clean, reads well (and I love the pure-function-components).

**Comment by ockham** [comment]
> ^ Potential fixes:
- Wrapping all the children of `Masterbar` in something neutral like a `div`
- Changing the type to `React.PropTypes.arrayOf( React.PropTypes.element )` and assuming multiple children will always be provided.
- Simply changing the type to `React.PropTypes.node`, which is more lax.

Opting for the latter. I'd hate adding a `div` here, since the Masterbar itself is only a `header` (cf earlier comments in this thread), and logged-out really only supplies one child.

**Comment by ockham** [comment]
Oh, you already fixed it :-D

File: client/layout/masterbar/logged-in.jsx [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+		this.setState( {
+			showNotifications: ! this.state.showNotifications
+		} );
+	},
+
+	isActive( section ) {
+		return section === this.props.section && ! this.state.showNotifications;
+	},
+
+	wordpressIcon() {
```
Comment (mcsf) [inline client/layout/masterbar/logged-in.jsx]: Totally nit-picky, and I understand this is not the scope of the refactor: should we use capital P?

**Comment by mcsf** [comment]
Regardless of the `wordpressIcon` question above, :+1: !

File: client/layout/masterbar/logged-in.jsx (hunk 2 above)
Comment (ockham) [inline client/layout/masterbar/logged-in.jsx]: nitPicky :-)

I'll respectfully stick with the lower-case p here. We shouldn't do `WordPressIcon()` for a method name, and IMO `wordPressIcon()` isn't any better than `wordpressIcon()`. So let's stick with the latter and succumb to `snakeCase`.

`</BikeShed>`

**Pull request merged**